### PR TITLE
Avoid unbound symbols in insignificant output strides

### LINKED ISSUE #148199: [Inductor-CPU] qlinear_binary output may have strides whose value can't be determined
### 🐛 Describe the bug

When `M` dimension is 1 and dynamic shape support is enabled with `torch.compile`, `qlinear_binary` op's output's outermost dim's stride may have a symbolic value that can't be determined.

Please copy-paste this UT in `test/inductor/test_mkldnn_pattern_matcher.py`:

```python
    @skipIfNoDynamoSupport
    @skipIfNoONEDNN
    # TODO: investigate options of torch.compile in fbcode
    @unittest.skipIf(IS_FBCODE, "Failing in fbcode")
    @parametrize("has_bias", [False,])
    @parametrize("dtype", [torch.bfloat16,])
    @parametrize("dynamic", [True,])
    @parametrize("reshape_a", [True,])
    @parametrize(
        "M",
        [
            1,
        ],
    )
    @parametrize("inplace_add", [True,])
    @parametrize("expand_a_scale", [False,])
    def test_reproduce_issue(
        self, has_bias, dtype, dynamic, reshape_a, M, inplace_add, expand_a_scale
    ):
        r"""
        This testcase check if we can match the int8_dynamic_activation_int8_weight int8 linear pattern from torchao,
        when activation is symmetrically quantized dynamically & weights are symmetrically quantized (statically)
        The pattern is:
            (no bias) _int_mm -> convert_element_type -> ([expand_a] -> mul) -> mul
        or
            (with bias) pattern_no_bias -> add
        Expansion of the scale of activation is optional.
        The pattern depiction doesn't mean that convert_element_type output is fed into expand_a as input,
        but simply that activation scale may be applied after an expand operation on it.
        """
        if dtype == torch.bfloat16 and not torch.ops.mkldnn._is_mkldnn_bf16_supported():
            return
        in_feature = 32
        out_feature = 64
        q_min, q_max = -32, 31
        # we only test for qlinear_binary in this case

        class Mod(torch.nn.Module):
            def __init__(self, dtype: torch.dtype, has_bias: bool):
                super().__init__()
                self.dtype = dtype
                self.has_bias = has_bias
                self.b = torch.randint(
                    q_min, q_max, [in_feature, out_feature], dtype=torch.int8
                )
                self.a_scale = torch.rand([M, 1], dtype=dtype) * 0.01 + 0.01
                self.b_scale = torch.rand([out_feature]) * 0.01 + 0.01
                self.b_scale = self.b_scale.to(dtype)
                self.bias = torch.rand([out_feature], dtype=dtype) if has_bias else None
                self.additive = torch.rand([M, out_feature], dtype=dtype)

            def forward(self, a):
                if reshape_a:
                    a_reshaped = a.reshape(-1, a.size(-1))
                else:
                    a_reshaped = a
                c = torch._int_mm(a_reshaped, self.b)
                c = c.to(self.dtype)
                if expand_a_scale:
                    a_scale = self.a_scale.expand(c.shape)
                else:
                    a_scale = self.a_scale
                c = c * a_scale
                c = c * self.b_scale
                if self.has_bias:
                    c = c + self.bias
                elif inplace_add:
                    # When M is 1, dynamic shapes are enabled with torch.compile, has_bias is False,
                    # expand_a_scale is False and inplace_add is true,
                    # the output's outermost dim's stride can't be determined due to some Dynamo bug.
                    c.add_(self.additive)
                return c

        mod = Mod(dtype, has_bias).eval()
        a = torch.randint(q_min, q_max, [M, in_feature], dtype=torch.int8)

        def matcher_check_fn():
            self.assertEqual(
                counters["inductor"]["qlinear_weight_prepack_matcher_count"], 1
            )

        self._test_common(
            mod,
            (a,),
            matcher_check_fn,
            check_autocast=dtype,
            compile_options={"dynamic": dynamic},
        )
```

### Versions

Main branch

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185781

Dynamic-shape output metadata can retain symbolic strides derived from frozen
module tensors. Freezing removes the runtime inputs for those symbols, while
output layout reconciliation can still copy the stale metadata into generated
code. Conversely, a valid stride symbol can be bound by a live graph input
without appearing in a newly allocated output layout. Treating layout
membership as codegen availability therefore admits orphaned symbols and
rejects valid dynamic strides.

Build a conservative set of symbols that wrapper codegen binds from live
scalar graph inputs, surviving symbolic input sources, and already-defined
unbacked symbols. Reconcile each empty or singleton dimension independently:
copy its requested stride only when every symbol is available. Keep current
significant strides when the layouts are guard-equivalent; when a flexible
layout changes order, use the requested significant strides while still
filtering unsafe insignificant ones.

Using live runtime bindings instead of ShapeEnv's global symbol set is
intentional because ShapeEnv retains symbols from inputs removed by freezing.
Fixing this in shared layout reconciliation covers both fixed and flexible
layouts without adding freezing-specific metadata handling.

Fixes #148199

Generated by my agent

Test Plan:

- `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_cpu_repro.py -k output_stride -v`
- `TORCHINDUCTOR_CPP_WRAPPER=1 TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_cpu_repro.py -k output_stride -v`
- `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_cpu_select_algorithm.py -k test_qlinear_binary_sum_dynamic_m1_output_stride -v`
- `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_torchinductor.py -k insignificant_strides -v`
- `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_torchinductor.py -k output_strides -v`
- `python test/inductor/test_torchinductor.py -k flexible_layout_immutable_free_symbols -v`
- `PATH="$PWD/.venv/bin:$PATH" lintrunner -a`
- `git diff --check && git diff --cached --check`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the regression after CI showed the aarch64 shard passing the compiled computation but reporting qlinear_binary_lower_count as 0. The test now still checks correctness, shape, and output stride on all CPU targets, and only asserts the qlinear_binary lowering counter on non-ARM64 targets where that lowering is expected. The docs failures are from the docs upload step missing aws in the runner image after docs generation, so they appear unrelated to this PR.

**Comment by jansel** [comment]
My agent says I rebased onto current main and updated the fix for the newer FlexibleLayout output path. The shared stride reconciliation now prevents orphaned frozen-shape symbols in both the FixedLayout/qlinear and FlexibleLayout exact-stride paths, with regressions for direct qlinear and m=0/1/2 _int_mm outputs. Focused tests and lintrunner -a pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says 

> The PR description is now synchronized with the current head: it documents the FlexibleLayout exact-stride path, the shared reconciliation logic, and the current qlinear plus m=0/1/2 regression coverage. No code changed and CI was not retriggered.

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -519,9 +519,32 @@ def significant_strides_equal(
...
+    """
+    available_symbols = free_symbols([*layout.size, *layout.stride, layout.offset])
+    new_stride = [*layout.stride]
+    is_empty = any(
+        V.graph.sizevars.statically_known_equals(dim, 0) for dim in layout.size
+    )
+    for i, size in enumerate(layout.size):
+        if is_empty or V.graph.sizevars.statically_known_leq(size, 1):
+            requested_symbols = free_symbols([strides[i]])
+            if requested_symbols.issubset(available_symbols):
```
Comment (bobrenjc93) [inline torch/_inductor/ir.py]: Presence in this output layout is not equivalent to availability in generated code. A symbolic stride can be bound from a graph input yet absent from a freshly allocated output layout; for example, cloning a dynamic dense `(1, N)` input with stride `(S, 1)` should preserve eager's `(S, 1)`, but this check rejects the bound `S` and returns `(N, 1)`. Check actual graph/runtime bindings while excluding symbols removed by freezing, and add a dynamic noncanonical singleton-stride regression.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -5007,7 +5026,10 @@ def as_stride_order(
     def as_exact_strides(
         self, exact_strides: Sequence[_IntLike], allow_padding: bool = False
     ) -> FixedLayout:
-        new_stride = exact_strides
+        if significant_strides_equal(exact_strides, self.stride, self.size):
+            new_stride = _match_insignificant_strides(self, exact_strides)
+        else:
+            new_stride = exact_strides
```
Comment (bobrenjc93) [inline torch/_inductor/ir.py]: When any significant stride differs, this branch copies the entire requested vector and bypasses the new safety check. For a size `(1, M, N)` layout whose requested strides are `(s_unbound, 1, M)` and whose realized significant order differs, `s_unbound` is still emitted and can produce the same NameError. Build the result per dimension: use requested significant strides, but independently filter unavailable strides on size-0/1 dimensions, with a regression covering differing significant strides.

Comment (jansel) [reply torch/_inductor/ir.py]: My agent says 

> Addressed in the updated head. `FlexibleLayout.as_exact_strides()` now reconciles every dimension even when significant strides differ: requested significant strides are retained, while an unavailable singleton or empty stride falls back to the natural layout. `test_exact_output_stride_filters_unbound_singleton_stride` reproduces the prior undefined-symbol path and verifies the safe mixed stride.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/ir.py]: My agent says 

> Addressed in the updated head. The stride matcher now checks live scalar graph inputs, surviving symbolic input sources, and already-defined unbacked symbols, so a dynamic input stride such as `S` is preserved even when it is absent from the new output layout. `test_dynamic_output_stride_preserves_bound_singleton` covers the noncanonical `(1, N)` clone case; local validation passed with both Python and C++ wrappers.

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -520,9 +520,66 @@ def significant_strides_equal(
     return True
 
 
+def _get_runtime_bound_symbols() -> OrderedSet[Symbol]:
+    """Return a conservative set of symbols bound by wrapper codegen."""
+    result = OrderedSet(V.graph.bound_unbacked_symbols)
```
Comment (mlazos) [inline torch/_inductor/ir.py:525]: _Reviewed by me; this finding applies here._

_AI-generated analysis (Codex):_

> **Unbacked symbols disappear when assertions are disabled**
> 
> When `do_not_emit_runtime_assertions=True`, operations still emit definitions for unbacked symbols, but `bound_unbacked_symbols` remains empty. This helper therefore rejects valid requested singleton strides and changes observable output metadata.
